initial sketch

EECOLOR opened the pull request:
This is a first sketch version. I mainly seek feedback for the example (the implementation is very rough and will change quite a bit).

Things I will would like to add for the first published version:

- a way to remove duplication of the double route paths (can already be achieved by creating consts in a separate file)
- (context aware) reverse routing
- (context aware) route tree, a route definition that allows you to define metadata and data fetching

EECOLOR committed:
initial sketch

EECOLOR committed:
remove rogue files

EECOLOR committed:
code cleanup

JoostKiens commented on example/src/App.js:
```diff
@@ -0,0 +1,75 @@
+  return (
+    <>
+      <Navigation />
+      {pick(
```
I feel `pick` might be a confusing name for this function (at least in a public API) since the name `pick` is commonly used for creating an object composed of picked object properties. 

https://www.npmjs.com/package/object.pick
https://lodash.com/docs/4.17.15#pick

EECOLOR replied:
Would `match` be better?

EECOLOR replied:
Or `routes`?

JoostKiens replied:
I like `match` or perhaps `matchRoute`?

JoostKiens commented on src/index.js:
```diff
@@ -0,0 +1,153 @@
+function Routing({ routes, initialLocation }) {
+  const context = React.useContext(locationContext)
+  if (!context && !inBrowser && !initialLocation)
+    throw new Error(`Your need to supply an initial location on server side rendering`)
```
There's a typo: "Your" should be "You".
Additionally, I feel like it would be more clear if it would explain what I would need to do:

`You'll need to add the "initialLocation" prop on the "LocationProvider" for server-side rendering`

Or something like that

JoostKiens commented on src/matching.js:
```diff
@@ -0,0 +1,59 @@
+  return typeof x === 'function' ? x(...args) : x
+}
+
+match.cache = {}
```
Just curious, but why is this a property of the function below instead of a separate object? Is it to make a connection between the function and the object? 

It feels a bit convoluted: I need to think twice before I understand what's going on.

EECOLOR replied:
Great question. Over the years it has gradually changed and apparently I ended up doing this. Originally it was like this:

```js
function test() {
  const cache = test.cache || (test.cache = {})
  if (test.cache && test.cache[key]) ...

  return test.cache[key] = ... 
}
```

Making the outside world oblivious to the fact that caching was used. This also makes it very portable.

The downside of this original approach is that typescript does not know about the `cache` property. The easiest route to keep the red squiggles at bay was to move `test.cache = {}` to it's separate line. The upside of this move is that it's instantly clear that this function has a form of caching, which has the following benefits:
- Can (almost) safely be called very often
- Returns stable values (handy when used in conjunction with hooks)

So all in all I think it is better to have `match.cache = {}` than `const matchCache = {}`

We could however think about introducing a utility (memoization), this however requires us to do this:

```js
export { 
  match: momoize(match)
}
```

And in some case (haven't checked this case) things become more problematic and less performant.

JoostKiens commented:
I like it 😀

peeke commented on src/index.js:
```diff
@@ -0,0 +1,153 @@
+}
+
+function resolve(basePath, to) {
+  return to.startsWith('/') ? to : `${basePath}/${to}`
```
Nice :+1:

peeke commented on src/index.js:
```diff
@@ -0,0 +1,153 @@
+function LinkImpl({ to, children }) {
+  const { basePath, navigate } = React.useContext(baseContext)
+  const href = resolve(basePath, to)
+  return <a {...{ href, children, onClick }} />
```
Might be useful to allow a className to be passed to this element, in case you want to create a larger hitarea using the `<a>`s `::before` of `::after`. Or would you say, just add a child which provides the hitarea?

EECOLOR replied:
Yeah, this one needs some work. I can imagine we do not provide a component so we can use our own (project specific) links. I'll check some projects to see how we normally handle links.

peeke reviewed (commented):
Looks great! 

One question: would it be useful to expose a `useMatch` hook? I've used it in a project with @reach/router where I wanted to match multiple routes, since `pick` matches exactly one route. 

The usecase was a map, which was shown for `/:param/*` and a separate overlay component, which was shown for `/:param/:slug/`:

```js
<div>
  <Component path='/:param/*' />
  <Drawer path='/:param/:slug' />
  <Overlay path='/:param/:slug' />
</div>
```

This way, it was clear what the specific routes where at the root level and I didn't have to wrap components in a wrapper component, just to match them to the same route. 

In some other routing libraries there's a difference between a router (displays any routes that match) and a switch (displays the first matching route). Maybe we could do something like `const { pick, matchAll } = useRouter({ initialLocation })`

Another solution would be to use nested routes, but I'm not sure that's objectively better. A little bit of context: this application was quite fluid, it was difficult to match a route to a single, traditional page component.

EECOLOR commented:
@peeke About the different types of matches. With the current setup you could do this:

```jsx
<div>
  {pick(['/:param/*', params => <Component {...params} />])}
  {pick(['/:param/:slug', params => <Drawer {...params}  />])}
  {pick(['/:param/:slug', params => <Overlay {...params}  />])}
</div>
```

@JoostKiens commented on the name of the method. I don't have a favorite yet. On top of that, we could make it a single overloaded method or two methods to support this use case better.

```jsx
  <div>
    {matchRoutes(
      ['/a', <a />], 
      ['/b', <b />],
    )}
    {matchRoute('/c', <c />)}
    {matchRoute('/c', <c />)}
  </div>
```
or
```jsx
  <div>
    {match(
      ['/a', <a />], 
      ['/b', <b />],
    )}
    {match('/c', <c />)}
    {match('/c', <c />)}
  </div>
```
or
```jsx
  <div>
    {routes(
      ['/a', <a />], 
      ['/b', <b />],
    )}
    {route('/c', <c />)}
    {route('/c', <c />)}
  </div>
```

Looking at this it seems the last one is my favorite.

Ohw, and your original version could be solved like this (using the `route` syntax) if you want to use nesting (at least I think):

```jsx
<div>
  {route(':param/*', params =>
    <>
      <Component {...params} />
      {route(':slug', params =>
        <>
          <Drawer {...params} />
          <Overlay {...params} />
        </>
      )}
    </>
  )}
</div>
```

EECOLOR commented:
Hmm, I think we can skip the `useRouting` for all non top-level uses completely. Although I don't think it's a good idea because it's harder to remember when to use what:

```jsx
import { useRouting } from '@kaliber/routing'

function Root({ initialLocation }) {
  const { routes } = useRouting({ initialLocation })
  return routes(
    ['a', <Child />],
    ['b', <Child />],
  )
}


import { routes } from '@kaliber/routing'

function Child() {
  return routes(
    ['a', <a />],
    ['b', <b />],
  )
}
```

vs

```jsx
import { useRouting } from '@kaliber/routing'

function Root({ initialLocation }) {
  const { routes } = useRouting({ initialLocation })
  return routes(
    ['a', <Child />],
    ['b', <Child />],
  )
}


import { useRouting } from '@kaliber/routing'

function Child() {
  const { routes } = useRouting()

  return routes(
    ['a', <a />],
    ['b', <b />],
  )
}
```

EECOLOR committed:
improve Link

EECOLOR committed:
improve code hinting during development

EECOLOR committed:
change `pick` to `routes`  and improve typing

EECOLOR committed:
add `route` to `useRouting`

peeke commented:
Can't wait to use this in production :)

> Looking at this it seems the last one is my favorite.

My favorite would be either the first or second one. From the last one it's not as clear that just one route is matched.

EECOLOR commented on src/index.js:
```diff
@@ -0,0 +1,202 @@
+    },
+    [basePath, routePath, navigate, params, baseParams]
+  )
+  const children = callOrReturn(createChildren, value.baseParams)
```
It would be nicer if this call was in the parent

EECOLOR commented on src/index.js:
```diff
@@ -0,0 +1,202 @@
+  return {
+    routes,
+    route(...route) { return routes(route) },
+  }
```
Can we add something that allows us to match routes to non-react stuff?

```js
const { match } = useRouting()
const x = match(['a', 1], ['b', 2])
```
Could be a good idea to move this to another hook

EECOLOR committed:
first sketch

please do not look the code

EECOLOR committed:
more sketching

Explored exhaustive /* routes based on children. It can be done. Most
likely candidate is making `extractPath` return an array of all possible
child routes (separated with /*). That requires a change in matching
where /*abcdefg is treated differently from /*. We probably also need
to change scoring to include length for equal scores (or use length as
a base of the score).

EECOLOR committed:
add relative pick

EECOLOR committed:
first sketch with reverse routing

EECOLOR committed:
more sketching, making the language switch work

EECOLOR committed:
Simplification of the library

- Route map is now a default (optional) integration
- We explicitly require LocationProvider to prevent 'magic'
- No more implicit LocationProviders

EECOLOR committed:
Update TODO's

EECOLOR committed:
Merge pull request #2 from kaliberjs/route-map

Route map

EECOLOR commented:
Will re-open a new pull request
